### ISSUE TITLE
v3 - Constraints don't appear to work as expected

### ISSUE THREAD

**Issue by antonosmond** [Masterminds/semver#150]
The semver spec states that prerelease versions always have a lower precedence than a normal version (https://semver.org/#spec-item-11).
Given:
```
1.0.0-rc1
1.0.0
```
`1.0.0-rc1` should be < `1.0.0`.
Given a constraint e.g. `< 1.0.0` I'd expect `1.0.0-rc1` to be flagged as true in the constraint check because it IS less than `1.0.0` however this doesn't appear to be working.
Have a look at this: https://play.golang.org/p/boXFhusKDBw
In every case, the prerelease version "should" have a value of `true` but it doesn't.
Is this a bug or am I misunderstanding something?

**Comment by antonosmond** [comment]
What's really strange is the final constraint check - `<=1.0-0`
In this check `1.0.0-rc1` passes the constraint as `true`. We know that `1.0.0-rc1` is NOT equal to `1.0.0` therefore it must be less than `1.0.0`. If this is true then the constraint `<1.0-0` should also flag it as true but it flags it as false.

**Comment by yanislavm** [comment]
Isn't this topic related to https://github.com/Masterminds/semver/issues/21

**Comment by mattfarina** [comment]
@antonosmond I think I can answer this. `<` is less than. So, `< 1.0.0-0` is saying that it's looking for versions that are less than `1.0.0-0`. The `-0` is saying to use `0` as the prerelease. SemVer uses Ascii characters so sorting can be funny. See https://www.asciitable.com/ for ordering. `0` is the lowest value accepted by the spec.

I modified your playground a little to show this. See https://play.golang.org/p/w4UElRmiFQz.
In your original ranges I tested a version of `0.9.0` as it's less than `1`. It shows true. I also created a range that's `>= 1.0.0-0`. This would look for versions that are greater than `1.0.0` and include all prereleases. It shows true for pre-releases.

I hope this helps.

**Comment by squaremo** [comment]
@mattfarina From the output of https://play.golang.org/p/w4UElRmiFQz:

```
Constraint: <1.0-0
  1.0.0-rc1: false
Constraint: <=1.0-0
  1.0.0-rc1: true
```

So 1.0.0-rc1 is not less than 1.0-0, but is less than or equal to 1.0-0, so ... it must be equal?

**Comment by mattfarina** [comment]
@squaremo I can answer this. `1.0-0` translates into `1.0.x-0`. The x is a wild card not a 0.

**Comment by mattolenik** [comment]
This seems like a pretty severe bug. `1.0.0-beta1` is not considered less than `2.0.0`!
https://play.golang.org/p/y8V58z5BGWx

**Comment by afirth** [comment]
@mattolenik expanding to https://play.golang.org/p/t8fAR9oVDV8
```
Constraint: >=1.0.0
  1.0.0-beta: false
Constraint: <1.0.0
  1.0.0-beta: false
Constraint: <1.x-0
  1.0.0-beta: false
Constraint: =1.x-0
  1.0.0-beta: true
Constraint: >1.x-0
  1.0.0-beta: false
```
it's _not compatible_ with ranges that don't specify they explicitly work with prereleases. Based on @mattfarina's explanation above, 1.x-0 translates to 1.x.x-x, all wildcards, and it's equal. See also https://github.com/Masterminds/semver/issues/21 and https://semver.org/#spec-item-9

honestly i think this could be closed, working as intended in https://github.com/Masterminds/semver/pull/23

**Comment by mattfarina** [comment]
It's important to note that semver has a specific format for versions. The range syntax that is in use has no specification. In fact, different systems do ranges differently. Helm is currently patterned after what JS and Rust do. The JS libs (used in npm itself) may be the most widely used range usage.

**Comment by mattolenik** [comment]
I must be missing something...how does any of that have bearing on a version starting with major version 1 being less than a version starting with 2? Should I open a second issue for that? Surely my example cannot be expected behavior!

**Comment by mattfarina** [comment]
To document, here's some details on what's going on with the examples...

`=1.x-0` can be a little hard to parse. It's equal to `1.x.x-0`. The `.x.x` is a range. So, this expands to `>= 1.0.0-0 < 2.0.0-0`. That means that `1.0.0-beta` is within that range. The `=` at the front is misleading because it equals a range.

`>1.x-0` expands to `> 1.0.0-0`. It's a range. `1.0.0-beta` fits within that range.

When wildcards are expanded to ranges it helps to explain why the results are what they are.

**Comment by skyzyx** [comment]
I know that this is old, but it's still open. Also, some of the Go Playgrounds differed from what was _actually being asked_, and according to the spec, @mattolenik and @antonosmond are correct that this is a bug.

Let's get rid of the `-0` nonsense that befuddled this discussion. Let's just use straight-up, normal versions and ranges.

https://go.dev/play/p/NbjUjgQHpwL

This Go Playground says that:

* (1.0.0-beta1 < 1.0.0) == false
* (1.0.0-beta1 < 2.0.0) == false

…which any human knows is incorrect. Reviewing https://semver.org/#spec-item-11, §11.1,2 says:

> 1. Precedence MUST be calculated by separating the version into major, minor, patch and pre-release identifiers in that order (Build metadata does not figure into precedence).
>
> 2. Precedence is determined by the first difference when comparing each of these identifiers from left to right as follows: Major, minor, and patch versions are always compared numerically.
>
> Example: 1.0.0 < 2.0.0 < 2.1.0 < 2.1.1.

We definitely see this in the example above with numeric-only versions. The following comparisons match both the spec and also human intelligence.

* (0.9.0 < 1.0.0) == true
* (1.0.0 < 1.0.0) == false
* (2.0.0 < 1.0.0) == false
* (0.9.0 < 2.0.0) == true
* (1.0.0 < 2.0.0) == true
* (2.0.0 < 2.0.0) == false

When we get to §11.3, however, we begin to see issues.

> 3. When major, minor, and patch are equal, a pre-release version has lower precedence than a normal version:
> 
> Example: 1.0.0-alpha < 1.0.0.

If we just look at the example provided, this package fails the test.

* (1.0.0-alpha < 1.0.0) == false **(but should be `true`)**

If we take the other obviously-prerelease versions, we see the following — all of which are _incorrect behavior_ per §11.1-3.

* (1.0.0-beta1 < 1.0.0) == false **(but should be `true`)**
* (1.0.0-rc1 < 1.0.0) == false **(but should be `true`)**
* (1.0.0-rc.1 < 1.0.0) == false **(but should be `true`)**

> [!IMPORTANT]
> I want to be absolutely clear that this is **not** comparing against `1.0.0-0` (prerelease). Instead, it's comparing against `1.0.0` (stable).

I understand that while versions like `1.0.0-1` are common in the real world (e.g., Linux packaging) and have their own specifications, these versions are not valid _Semantic Versions_ (per its spec). If we look at §11.4, this is explained.

> 4. Precedence for two pre-release versions with the same major, minor, and patch version MUST be determined by comparing each dot separated identifier from left to right until a difference is found as follows:
>
>     1. Identifiers consisting of only digits are compared numerically.
>
>     2. Identifiers with letters or hyphens are compared lexically in ASCII sort order.
>
>     3. Numeric identifiers always have lower precedence than non-numeric identifiers.
>
>     4. A larger set of pre-release fields has a higher precedence than a smaller set, if all of the preceding identifiers are equal.
>
> Example: 1.0.0-alpha < 1.0.0-alpha.1 < 1.0.0-alpha.beta < 1.0.0-beta < 1.0.0-beta.2 < 1.0.0-beta.11 < 1.0.0-rc.1 < 1.0.0.

However, I want to explicitly call out this part — which _implies_ that this rule also applies to _versions with different major, minor, and patch values_ (although does not directly state it). There is a better explanation in §2, where it says:

> 2. A normal version number MUST take the form X.Y.Z where X, Y, and Z are non-negative integers, and MUST NOT contain leading zeroes. X is the major version, Y is the minor version, and Z is the patch version. Each element MUST increase numerically. For instance: 1.9.0 -> 1.10.0 -> 1.11.0.

Rules are applied _in-order_:

1. We go left to right, comparing along the way.
2. Identifiers consisting of only digits are compared numerically.
3. Once we get to a prerelease, we compare lexicographically by US-ASCII bytecode.

So, when we use this package to compare certain versions, we can see that the package violates the rules of both §2 and also §11.4.

* (1.0.0-alpha < 2.0.0) == false **(should be `true`)**
* (1.0.0-beta1 < 2.0.0) == false **(should be `true`)**
* (1.0.0-rc1 < 2.0.0) == false **(should be `true`)**
* (1.0.0-rc.1 < 2.0.0) == false **(should be `true`)**

What the spec says, and also what humans expect, is the following:

* (1.0.0-rc.1 < 2.0.0) == true (package says `false`)
* (1.0.0-rc < 2.0.0) == true (package says `false`)
* (1.0.0 < 2.0.0) == true (package says `true`)
* (1.0 < 2.0) == true (package says `true`)
* (1 < 2) == true (package says `true`)

Looking at a second example, https://go.dev/play/p/XOHvsLysvWE, I threw in a comparison to a prerelease. And when you do this, the comparison is suddenly correct.

* (1.0.0-alpha < 2.0.0-rc.1) == true **(should be `true`)**
* (1.0.0-beta1 < 2.0.0-rc.1) == true **(should be `true`)**
* (1.0.0-rc1 < 2.0.0-rc.1) == true **(should be `true`)**
* (1.0.0-rc.1 < 2.0.0-rc.1) == true **(should be `true`)**
* (1.0.0-alpha < 2.0.0-0) == true **(should be `true`)**
* (1.0.0-beta1 < 2.0.0-0) == true **(should be `true`)**
* (1.0.0-rc1 < 2.0.0-0) == true **(should be `true`)**
* (1.0.0-rc.1 < 2.0.0-0) == true **(should be `true`)**

There appears to be a bug when comparing prerelease vs non-prerelease versions where (at the very least), comparing major versions stops producing correct results.

Of final note is that sorting appears to work correctly. https://go.dev/play/p/hLAs7QTgVln. This bug seems to only affect comparators.

**Comment by mattfarina** [comment]
There is documentation on handling pre-releases that is relevant to this thread. See https://github.com/masterminds/semver?tab=readme-ov-file#working-with-prerelease-versions

It explains how pre-release handling works, why is works that why, and why what's here is causing the results it is.

To make a change from the current behavior would be a breaking change which would require a v4 major version of the library in order to make.

Libraries in other languages handle pre-releases as special cases, too. For example, see what node-semver currently does... https://github.com/npm/node-semver?tab=readme-ov-file#prerelease-tags. It also requires you to opt-in to pre-release handling.

Since the spec is being quoted...

> A pre-release version indicates that the version is unstable and might not satisfy the intended compatibility requirements as denoted by its associated normal version.

To quote node-semver on the topic...

> The purpose of this behavior is twofold. First, prerelease versions frequently are updated very quickly, and contain many breaking changes that are (by the author's design) not yet fit for public consumption. Therefore, by default, they are excluded from range-matching semantics.

The _why_ here is important.

**Comment by mattolenik** [comment]
How is it a breaking change to make it work as intended? It's very obvious this is broken and examples have been shown.

I really don't understand the pushback here.

**Comment by mattfarina** [comment]
First, pre-releases are unstable and might not satisfy the intended compatibility requirements of the normal stable versions. The spec explicitly states this. Early versions of this library treated pre-releases the same as releases and it turned out to be a problem. People using tools built on this library accidentally deployed pre-release software where they expected to deploy the stable version. The current functionality was born out of experience.

Second, the spec states precedence with pre-releases and that pre-releases may not provide compatibility. There is no spec for tools that do comparisons. Since pre-releases may not be compatible (and I have seen cases where they aren't) they are not included by default. You need to opt-in to them.

This semver package is not unique in leaving pre-releases out by default. In my previous comment I showed how node-semver (used by npm) requires opt-in. Rust provides another example of a place where you need to opt-in to pre-releases.

Third, you're telling the author of this library (who has read implementations in other languages, has read popular tools that use the libraries across languages, and dealt with many situations from users of this library) that he's wrong on _intent_ while leaving out the pre-release incompatibility part. I've been developing this library for almost 9 years.

I'm not saying there aren't API changes I would make to this library if a v4 comes up. But, the case for pre-release being on-by-default is going to need a good case. 

Why are these (multiple listed here) projects wrong for opting in to pre-release?

**Comment by skyzyx** [comment]
We're exclusively talking about _comparing_ versions. And when _comparing_ versions, my comment above still stands.

We're not talking about _guarantees_; we're talking about comparing version numbers.

If you don't want to fix it, just say so. But @mattolenik and my logic is sound. The software does not behave as users expect, and arguably as the spec defines.

**Comment by mattfarina** [comment]
Note, there is no spec for constraint systems. Using the compare features on the `Version` type deals with pre-releases all of the time. This is compliant to the semver spec.

Constraint systems (e.g., when you parse a string with characters like `^`, `~`, etc) do **_not_** have a spec. If you look at constraint systems in Go, JS, Rust, and others you'll find differences in the way things are handled. One thing most of these systems have in common is to opt-in to pre-releases. This is due to practical considerations in real world use and because pre-releases are not expected to meet the same stability guarantees as stable versions.

A big thing here is to look at intent. Pre-releases are not expected to meet the same stability guarantees as stable releases.

If there was an issue with the compare functionality, that would be a bug. There is no issue with the intended design of the constraint system.
